lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/api/DataLoader.java

```diff
@@ -97,33 +144,52 @@ public void onNext(final Alarm time) {
 
     this.requestor = requestor;
     this.dataLoadingService = dataLoadingService;
-    this.dataEvalMemoryMB = dataEvalMemoryMB;
-    this.dataEvalCore = dataEvalCore;
     this.resourceRequestHandler = new ResourceRequestHandler(requestor);
-    this.resourceRequestStage = new SingleThreadStage<>(this.resourceRequestHandler, 2);
-
-    if (serializedComputeRequest.equals("NULL")) {
-      this.computeRequest = null;
+    // the resource request queue will have as many requests as compute and data requests.
+    this.resourceRequestStage = new SingleThreadStage<>(
+        this.resourceRequestHandler, serializedComputeRequests.size()
+            + serializedDataRequests.size());
+
+    // check if the size == 1 and if equals to NULL to maintain previous functionality
+    // TODO once the other constructor is completely removed, we should just check for emptiness
+    if (serializedComputeRequests.isEmpty()
+        || serializedComputeRequests.size() == 1
+        && NULL.equals(serializedComputeRequests.iterator().next())) {
       this.computeEvalMemoryMB = -1;
-      computeEvalCore = 1;
+      this.computeEvalCore = 1;
     } else {
-      this.computeRequest = EvaluatorRequestSerializer.deserialize(serializedComputeRequest);
-      this.computeEvalMemoryMB = this.computeRequest.getMegaBytes();
-      this.computeEvalCore = this.computeRequest.getNumberOfCores();
-      this.numComputeRequestsToSubmit.set(this.computeRequest.getNumber());
-
-      this.resourceRequestStage.onNext(this.computeRequest);
+      // Deserialize each compute request.
+      // Keep the maximum number of cores and memory requested, in case some
+      // evaluator fails, we will try to reallocate based on that.
+      for (final String serializedComputeRequest : serializedComputeRequests) {
+        final EvaluatorRequest computeRequest = EvaluatorRequestSerializer.deserialize(serializedComputeRequest);
+        this.numComputeRequestsToSubmit.addAndGet(computeRequest.getNumber());
+        this.computeEvalMemoryMB = Math.max(this.computeEvalMemoryMB, computeRequest.getMegaBytes());
+        this.computeEvalCore = Math.max(this.computeEvalCore, computeRequest.getNumberOfCores());
+        this.resourceRequestStage.onNext(computeRequest);
+      }
+    }
+    // Deserialize each data requests.
+    // We distribute the partitions evenly accross the DCs.
+    // The number of partitions extracted from the dataLoadingService override
+    // the number of evaluators requested (this preserves previous functionality)
+    final int dcs = serializedDataRequests.size();
+    final int partitionsPerDataCenter = this.dataLoadingService.getNumberOfPartitions() / dcs;
+    int missing = this.dataLoadingService.getNumberOfPartitions() % dcs;
+    for (final String serializedDataRequest : serializedDataRequests) {
+      EvaluatorRequest dataRequest = EvaluatorRequestSerializer.deserialize(serializedDataRequest);
```

Comment:
Nope, I'm reassigning it because I need to update the number of evaluators. The number is taken from the numberOfPartitions of the dataLoadingService, I maintained previous functionality.